File: src/lib/calc.ts
import type { Balance, Expense, Member, Settlement, Transfer } from "./types";

/**
 * §7.1 Equal split: floor division, remainder handed out one minor unit at a
 * time in ascending member position. Deterministic by construction.
 */
export function splitEqual(
  totalCents: number,
  participants: { id: string; position: number }[],
): { member_id: string; amount_cents: number }[] {
  const ordered = [...participants].sort((a, b) => a.position - b.position);
  const n = ordered.length;
  if (n === 0) return [];
  const base = Math.floor(totalCents / n);
  const remainder = totalCents - base * n;
  return ordered.map((p, index) => ({
    member_id: p.id,
    amount_cents: base + (index < remainder ? 1 : 0),
  }));
}

/** §7.2 net = paid − owed + settlements paid − settlements received */
export function computeBalances(
  members: Member[],
  expenses: Expense[],
  settlements: Settlement[],
): Balance[] {
  const net = new Map<string, number>();
  for (const m of members) net.set(m.id, 0);
  const bump = (id: string, delta: number) => {
    if (!net.has(id)) return;
    net.set(id, (net.get(id) ?? 0) + delta);
  };

  for (const expense of expenses) {
    bump(expense.payer_member_id, expense.amount_cents);
    for (const share of expense.shares) bump(share.member_id, -share.amount_cents);
  }
  for (const s of settlements) {
    bump(s.from_member_id, s.amount_cents);
    bump(s.to_member_id, -s.amount_cents);
  }

  return members.map((m) => ({ member_id: m.id, net_cents: net.get(m.id) ?? 0 }));
}

/** §7.3 Greedy largest-debtor → largest-creditor, tie-broken by position. */
export function simplifyTransfers(balances: Balance[], members: Member[]): Transfer[] {
  const position = new Map(members.map((m) => [m.id, m.position]));
  const order = (
    a: { member_id: string; amount: number },
    b: { member_id: string; amount: number },
  ) =>
    b.amount - a.amount ||
    (position.get(a.member_id) ?? 0) - (position.get(b.member_id) ?? 0);

  const debtors = balances
    .filter((b) => b.net_cents < 0)
    .map((b) => ({ member_id: b.member_id, amount: -b.net_cents }))
    .sort(order);
  const creditors = balances
    .filter((b) => b.net_cents > 0)
    .map((b) => ({ member_id: b.member_id, amount: b.net_cents }))
    .sort(order);

  const transfers: Transfer[] = [];
  let d = 0;
  let c = 0;
  while (d < debtors.length && c < creditors.length) {
    const debtor = debtors[d];
    const creditor = creditors[c];
    const amount = Math.min(debtor.amount, creditor.amount);
    if (amount > 0) {
      transfers.push({
        from_member_id: debtor.member_id,
        to_member_id: creditor.member_id,
        amount_cents: amount,
      });
    }
    debtor.amount -= amount;
    creditor.amount -= amount;
    if (debtor.amount === 0) d += 1;
    if (creditor.amount === 0) c += 1;
  }
  return transfers;
}

/**
 * §7.4 Raw pairwise: derived from expense shares alone, then settlements
 * applied between the same two members. Never reads the greedy output.
 */
export function computePairwise(
  members: Member[],
  expenses: Expense[],
  settlements: Settlement[],
): Transfer[] {
  const owes = new Map<string, number>();
  const key = (from: string, to: string) => `${from}|${to}`;
  const add = (from: string, to: string, amount: number) => {
    if (from === to) return;
    owes.set(key(from, to), (owes.get(key(from, to)) ?? 0) + amount);
  };

  for (const expense of expenses) {
    for (const share of expense.shares) {
      add(share.member_id, expense.payer_member_id, share.amount_cents);
    }
  }
  for (const s of settlements) add(s.to_member_id, s.from_member_id, s.amount_cents);

  const ordered = [...members].sort((a, b) => a.position - b.position);
  const result: Transfer[] = [];
  for (let i = 0; i < ordered.length; i += 1) {
    for (let j = i + 1; j < ordered.length; j += 1) {
      const a = ordered[i].id;
      const b = ordered[j].id;
      const netted = (owes.get(key(a, b)) ?? 0) - (owes.get(key(b, a)) ?? 0);
      if (netted > 0) {
        result.push({ from_member_id: a, to_member_id: b, amount_cents: netted });
      } else if (netted < 0) {
        result.push({ from_member_id: b, to_member_id: a, amount_cents: -netted });
      }
    }
  }
  return result;
}

File: backend/app/calc.py
"""Split and balance maths from spec §7. Pure functions over integer cents.

These rules must produce exactly the same numbers as the frontend's
`src/domain/calc.ts`, which the mock API used.
"""

from collections.abc import Hashable, Iterable, Sequence
from typing import NamedTuple

from app.models import ExpenseRecord, MemberRecord, SettlementRecord

type MemberKey = Hashable


class Participant(NamedTuple):
    member_id: MemberKey
    position: int


class NetBalance(NamedTuple):
    member_id: MemberKey
    position: int
    net_cents: int


class Transfer(NamedTuple):
    from_member_id: MemberKey
    to_member_id: MemberKey
    amount_cents: int


def split_equal(
    total_cents: int, participants: Iterable[Participant]
) -> list[tuple[MemberKey, int]]:
    """§7.1: floor division, then the remainder goes one minor unit at a time to
    participants in ascending position. Returned in position order."""
    ordered = sorted(participants, key=lambda p: p.position)
    if not ordered:
        return []
    base, remainder = divmod(total_cents, len(ordered))
    return [(p.member_id, base + (1 if i < remainder else 0)) for i, p in enumerate(ordered)]


def compute_balances(
    members: Sequence[MemberRecord],
    expenses: Iterable[ExpenseRecord],
    settlements: Iterable[SettlementRecord],
) -> list[NetBalance]:
    """§7.2: net = paid − own shares + settlements paid − settlements received.

    One entry per member, in ascending position, zeros included."""
    net = {m.id: 0 for m in members}

    def bump(member_id: MemberKey, delta: int) -> None:
        if member_id in net:
            net[member_id] += delta

    for expense in expenses:
        bump(expense.payer_member_id, expense.amount_cents)
        for share in expense.shares:
            bump(share.member_id, -share.amount_cents)
    for settlement in settlements:
        bump(settlement.from_member_id, settlement.amount_cents)
        bump(settlement.to_member_id, -settlement.amount_cents)

    ordered = sorted(members, key=lambda m: m.position)
    return [NetBalance(m.id, m.position, net[m.id]) for m in ordered]


def simplify_transfers(balances: Iterable[NetBalance]) -> list[Transfer]:
    """§7.3: repeatedly match the largest debtor with the largest creditor.

    Both sides are sorted by absolute amount descending, ties broken by
    ascending position, which makes the output deterministic."""
    entries = list(balances)

    def order(b: NetBalance) -> tuple[int, int]:
        return (-abs(b.net_cents), b.position)

    debtors = [[b.member_id, -b.net_cents] for b in sorted(entries, key=order) if b.net_cents < 0]
    creditors = [[b.member_id, b.net_cents] for b in sorted(entries, key=order) if b.net_cents > 0]

    transfers: list[Transfer] = []
    d = c = 0
    while d < len(debtors) and c < len(creditors):
        debtor, creditor = debtors[d], creditors[c]
        amount = min(debtor[1], creditor[1])
        transfers.append(Transfer(debtor[0], creditor[0], amount))
        debtor[1] -= amount
        creditor[1] -= amount
        if debtor[1] == 0:
            d += 1
        if creditor[1] == 0:
            c += 1
    return transfers


def compute_pairwise(
    members: Sequence[MemberRecord],
    expenses: Iterable[ExpenseRecord],
    settlements: Iterable[SettlementRecord],
) -> list[Transfer]:
    """§7.4: raw debts between each pair, from share and settlement rows only.

    Every participant owes their share to the payer; a settlement from A to B
    reduces what A owes B. Reciprocal debts are netted, zero pairs dropped, and
    pairs are listed in position order."""
    owed: dict[tuple[MemberKey, MemberKey], int] = {}

    def add(debtor: MemberKey, creditor: MemberKey, amount: int) -> None:
        if debtor != creditor:
            owed[(debtor, creditor)] = owed.get((debtor, creditor), 0) + amount

    for expense in expenses:
        for share in expense.shares:
            add(share.member_id, expense.payer_member_id, share.amount_cents)
    for settlement in settlements:
        add(settlement.to_member_id, settlement.from_member_id, settlement.amount_cents)

    ordered = sorted(members, key=lambda m: m.position)
    result: list[Transfer] = []
    for i, a in enumerate(ordered):
        for b in ordered[i + 1 :]:
            net = owed.get((a.id, b.id), 0) - owed.get((b.id, a.id), 0)
            if net > 0:
                result.append(Transfer(a.id, b.id, net))
            elif net < 0:
                result.append(Transfer(b.id, a.id, -net))
    return result
